langchain/agents/agent_toolkits/openapi/planner.py

```diff
@@ -0,0 +1,375 @@
+"""Agent that interacts with OpenAPI spec'd APIs via a hierarchical planning approach,
+inspired by recent works on LLMs for robotics application.
+
+Code's dumped here so it's self-contained while I'm experimenting.
+"""
+from langchain.agents.agent_toolkits.openapi.spec import (
+    reduce_openapi_spec,
+    ReducedOpenAPISpec,
+)
+from langchain.agents.agent import AgentExecutor
+from langchain.agents.mrkl.base import ZeroShotAgent
+from langchain.agents.tools import Tool
+from langchain.chains.llm import LLMChain
+from langchain.llms.openai import OpenAI
+from langchain.prompts import PromptTemplate
+from langchain.requests import RequestsWrapper
+from langchain.schema import BaseLanguageModel
+from langchain.tools.base import BaseTool
+from langchain.tools.requests.tool import BaseRequestsTool
+
+import json
+import re
+import yaml
+from typing import Optional
+
+
+# gpt-4 works very well with this approach, and the advantage is separating computation so as not to eat too many tokens.
+# gpt-3 completion models aren't as robust, but this approach allows for replanning at plan level + retrying at policy/step level.
+# I've not tried gpt-3 chat models as of yet.
+DEFAULT_LLM = OpenAI(model_name="gpt-4", temperature=0.0)
+
+
+#
+# Prompts.
+#
+
+# Planning.
+# TODO: remove API-specific few shots.
+API_PLANNER_PROMPT = """You are a planner that plans a sequence of API calls to assist with user queries against an API.  You have access to GET and POST endpoints, documented below.
```

Comment:
* What are your thoughts on adding DELETE / PATCH / PUT? 
* Do we need to add safeguards for operations that involve mutation?

Reply:
> What are your thoughts on adding DELETE / PATCH / PUT?
👍 for sure. I guess for now, this PR is really just demonstrative. We have to get this approach working a bit better for gpt3 family. Once that's there, I think extending to these other verbs is reasonable.


> Do we need to add safeguards for operations that involve mutation?
Yes definitely. Don't think there's much of a mechanism for this atm, but know some refactoring of agent execution is wip